### DISCUSSION TITLE
mirror physical storage

### DISCUSSION THREAD

**Discussion by hannesm** [openzfs/zfs#16285] [Q&A] [answered]
Dear everyone,

I have a ZFS mirror with two disks. Unfortunately, both disks had hardware issues at the same time. Now, I could read parts from each disk, and I'm wondering how a zfs mirror stores data, is it the case that sector N on disk 1 has the same contents as sector N on disk 2?

If this is the case, this means when I can read sectors A and C from disk 1 and sector B from disk 2, I could merge them as A[1], B[2], C[1] -- and get out more data than reading only disk 1 or only disk 2.

Does this make sense?

Thanks a lot for your reply,

Hannes

**Comment by amotin** [comment, marked as answer]
> is it the case that sector N on disk 1 has the same contents as sector N on disk 2?

Mostly yes.  Disks in ZFS mirror are identical except labels at the beginning and end of each disk, which are unique.  To use each disk ZFS needs at least one of its labels to be readable.  After that, if there are several readable copies of some data block block, unlike traditional RAIDs, thanks to its checksums ZFS can actually say which of them is the right one in each case.
